CenterForOpenScience/osf.io
# [ENG-2122] NodeStorage API Endpoint

## Purpose
Adds a node storage endpoint to access the storage usage (on OSFStorage) and storage limit status of a node and disallows a public project from being converted to a private project if it exceeds private storage limits.

## Changes
* Adds a `formatted_storage_usage` property to the Node model
* Adds checks to ensure the `storage_usage` property exists before calculating dependent properties
* Creates a NodeStorage API Endpoint
  * Adds NodeStorage relationship to NodeSerializer (creating an embeddable relationship)
  * Adds a WriteAdmin permission for any actions
* Ensures private storage limits are respected before converting a public node into a private node

## QA Notes

Please make verification statements inspired by your code and what your code touches.
- Verify the NodeStorage endpoint (`/v2/nodes/<node_id>/storage`) works for those with Write/Admin rights to the Node
  - If the `storage_usage` or `storage_limit_status` is not available, the results will be `null` with a `202` status code. If they are available, the results will be available with a `200` status code.
- Verify the `storage` endpoint is available on NodeDetail endpoint
  - Verify `?embed=storage` works on NodeDetail endpoints
- Verify that public projects exceeding private storage usage limits cannot be converted to private projects and that public projects under private storage usage limits may be converted.

What are the areas of risk?

Any concerns/considerations/questions that development raised?

## Documentation
API Documentation should be added for the new NodeStorage endpoint

## Side Effects
None Anticipated

## Ticket
[JIRA Ticket](https://openscience.atlassian.net/browse/ENG-2122)
